power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +366,550 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
+            std::invocable<std::remove_cvref_t<StateUpdater_>, ConstDataset const&> &&
+            requires(TransformerRanker_ const& ranker, State_ const& state) {
+                { ranker(state) } -> std::convertible_to<RankedTransformerGroups>;
+            }
+class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator, StateUpdater_, State_,
+                               TransformerRanker_> : public detail::BaseOptimizer<StateCalculator, State_> {
   public:
     using Base = detail::BaseOptimizer<StateCalculator, State_>;
     using typename Base::Calculator;
     using typename Base::ResultType;
     using typename Base::State;
     using StateUpdater = StateUpdater_;
+    using TransformerRanker = TransformerRanker_;
+
+  private:
+    using ComponentContainer = typename State::ComponentContainer;
+    using RegulatedTransformer = TapRegulatorRef<TransformerTypes...>;
+    using UpdateBuffer = std::tuple<std::vector<typename TransformerTypes::UpdateType>...>;
 
-    TapPositionOptimizer(Calculator calculator, StateUpdater updater, OptimizerStrategy strategy)
+    template <transformer_c T>
+    static constexpr auto transformer_index_of = container_impl::get_cls_pos_v<T, TransformerTypes...>;
+    static_assert(((transformer_index_of<TransformerTypes> < sizeof...(TransformerTypes)) && ...));
+
+  public:
+    TapPositionOptimizerImpl(Calculator calculator, StateUpdater updater, OptimizerStrategy strategy)
         : calculate_{std::move(calculator)}, update_{std::move(updater)}, strategy_{strategy} {}
 
-    auto optimize(State const& state) -> ResultType final {
-        auto const order = rank_transformers(state);
-        return optimize(state, order);
+    auto optimize(State const& state, CalculationMethod method) -> ResultType final {
+        auto const order = regulator_mapping<TransformerTypes...>(state, TransformerRanker{}(state));
+
+        auto const cache = this->cache_states(order);
+        auto optimal_output = optimize(state, order, method);
+        update_state(cache);
+
+        return optimal_output;
     }
 
-    constexpr auto get_strategy() { return strategy_; }
+    constexpr auto get_strategy() const { return strategy_; }
 
   private:
-    auto optimize(State const& /*state*/, RankedTransformerGroups const& /*order*/) -> ResultType {
-        // TODO(mgovers): implement outter loop tap changer
-        throw PowerGridError{};
+    auto optimize(State const& state, std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                  CalculationMethod method) const -> ResultType {
+        initialize(regulator_order);
+
+        if (auto result = try_calculation_with_regulation(state, regulator_order, method);
+            strategy_ == OptimizerStrategy::any) {
+            return result;
+        }
+
+        // refine solution
+        step_all(regulator_order);
+        return try_calculation_with_regulation(state, regulator_order, method);
```

Comment:
How could this work? 
(Let's first not discuss the fact that `::any` strategy is isolated and that if the `try_calc_` call fails to return anything meaningful, the subsequent runs should also be skipped because something is not right.)

Shouldn't there be an outer loop over like the following sketch?

```suggestion
        res{};
        while (...) {
                step_all(regulator_order); // the ::any check should happen here?
                res = try_calculation_with_regulation(state, regulator_order, method);
        }
        return res;
```

Reply:
i don't really understand what you mean.

the `try_calculation_with_regulation` contains the actual loop to saturate in a valid set of tap positions. However, that solution might not be optimal yet. To achieve an optimal solution, one more run is necessary where you re-try with one tap position lower for all transformers and re-run the optimization.

This logic was written by Tony. I'm fine with changing it, but in that case, also the outer loop document should be reviewed and updated accordingly.

Reply:
Yes, you're correct that I missed one line last Friday (reading code on github is not 100% optimal for me ;) ). It is actually the function name that misled me into understanding the `try_calculation_with_regulation` function call being a unit calculation call. Whereas in reality, it is the optimization iteration core. In that sense, I correct my second part of the comment to suggesting a function renaming.

The first part, however, is still up for discussion in my view.

Reply:
about your first part:

* `any` is isolated because **any** valid output of `try_calculation_with_regulation` is valid; no finetuning is necessary to achieve a more optimal state
* if `try_calculation_with_regulation` fails, then it will throw an error (e.g. `IterationDiverge` or `SparseMatrixError`). no additional check is necessary

about your second part:

* i get that `try_calculation_with_regulation` is a little bit cryptic but don't really have a better name. do you?

Reply:
> about your first part:
> 
> * `any` is isolated because **any** valid output of `try_calculation_with_regulation` is valid; no finetuning is necessary to achieve a more optimal state

This is due to the design that requires a refinement step, which is totally avoidable. I will attach a short pseudo code at the end of this comment/OneNote page as well as bring that up during the stand-up for discussion

> * if `try_calculation_with_regulation` fails, then it will throw an error (e.g. `IterationDiverge` or `SparseMatrixError`). no additional check is necessary
> 

In this case, why not simply put the actual logic in a try block instead of doing this abstraction?

> about your second part:
> 
> * i get that `try_calculation_with_regulation` is a little bit cryptic but don't really have a better name. do you?

I would name this `optimize` or `iterate` and instead call the existing `opt...` simply `run` or `run_opt`.


```python
# NB. extremely inaccurate use of names and types, which are not the point here
update_data = [{}*len(regulator_order)]
res = {}
last_iter_check = [False*len(regulator_order)]
while():
    tap_changed = False
    propose_data = {}
    while(rank < len(regulator_order)):
        same_rank_regulators = regulator_order[rank]:
        tap_changed = control_transformers(opt_strategy, same_rank_regulators, ..., res, propose_data)
        
        if not tap_changed and (opt_strategy == 'any' or last_iter_check[rank]):
            update_state(..., update_data)
            rank += 1
            continue # go to next rank

        if tap_changed or !last_iter_check[rank]:
            update_state(..., propose_data)
            res = calc_(...)

        last_run_check[rank] = !tap_changed
        update_data = update_data if last_iter_check[rank] else propose_data
        continue # still do current rank

return res
```

Reply:
> This is due to the design that requires a refinement step, which is totally avoidable. I will attach a short pseudo code at the end of this comment/OneNote page as well as bring that up during the stand-up for discussion

100% agree but IMHO out of scope of this PR. it's already long enough as is and this may be a future optimization

> In this case, why not simply put the actual logic in a try block instead of doing this abstraction?

actual errors are real errors. the optimizer should throw - not catch. no `try` block necessary. the `try_*` was just an unfortunate choice of words because of its ambiguity. see below.

> I would name this optimize or iterate and instead call the existing opt... simply run or run_opt.

i prefer `iterate` and keep `optimize` as is. if not, then maybe `find_local_optimum`. the `optimize` function would then specifically request a global optimum. that way, we could even go as far as to make the `local_minimum` and `local_maximum` strategies distinct from the `global_*` ones.

Reply:
Let's park this discussion for later.